core/include/detray/propagator/actors/event_writer.hpp

```diff
@@ -0,0 +1,166 @@
+}
+
+template <typename scalar_t>
+struct measurement_smearer {
+
+    measurement_smearer(scalar_t stddev_local0, scalar_t stddev_local1)
+        : stddev({stddev_local0, stddev_local1}) {}
+
+    measurement_smearer(measurement_smearer<scalar_t>& smearer)
+        : stddev(smearer.stddev) {}
+
+    std::array<scalar_t, 2> stddev;
+
+    std::random_device rd{};
+    std::mt19937 generator{rd()};
+
+    template <int ID>
+    scalar_t get() {
+        static_assert(ID == 0 || ID == 1);
+
+        if constexpr (ID == 0) {
+            return std::normal_distribution<scalar_t>(0, stddev[0])(generator);
+        } else if (ID == 1) {
+            return std::normal_distribution<scalar_t>(0, stddev[1])(generator);
+        }
+    }
+};
+
+template <typename transform3_t, typename smearer_t>
+struct event_writer : actor {
+
+    struct state {
+        state(std::size_t event_id, smearer_t smearer)
+            : m_particle_writer(get_event_filename(event_id, "-particles.csv")),
+              m_hit_writer(get_event_filename(event_id, "-hits.csv")),
+              m_meas_writer(get_event_filename(event_id, "-measurements.csv")),
+              m_meas_smearer(smearer) {}
+
+        std::size_t particle_id = 0;
+        particle_writer m_particle_writer;
+        hit_writer m_hit_writer;
+        measurement_writer m_meas_writer;
+        smearer_t m_meas_smearer;
+
+        void write_particle(const free_track_parameters<transform3_t>& track) {
+            csv_particle particle;
+            const auto pos = track.pos();
+            const auto mom = track.mom();
+
+            particle.particle_id = particle_id;
+            particle.vx = pos[0];
+            particle.vy = pos[1];
+            particle.vz = pos[2];
+            particle.vt = track.time();
+            particle.px = mom[0];
+            particle.py = mom[1];
+            particle.pz = mom[2];
+            particle.q = track.charge();
+
+            m_particle_writer.append(particle);
+            particle_id++;
+        }
+    };
+
+    struct measurement_kernel {
+        using output_type = point2;
+
+        template <typename mask_group_t, typename surface_t>
+        inline output_type operator()(
+            const mask_group_t& mask_group, const surface_t& surface,
+            const bound_track_parameters<transform3_t>& bound_params,
+            smearer_t smearer) {
+
+            const auto& mask = mask_group[surface.mask_range()];
+
+            auto local_coordinate = mask.local();
+
+            return local_coordinate.get_measurement(bound_params, smearer);
+        }
+    };
+
+    template <typename propagator_state_t>
+    void operator()(state& writer_state,
+                    propagator_state_t& propagation) const {
+
+        auto& navigation = propagation._navigation;
+        auto& stepping = propagation._stepping;
+
+        // @note: This should be changed for active surface rather than module
+        if (navigation.is_on_module()) {
```

Comment:
@niermann999 In this event writer, I want to record measurement only on the active planes, but `is_on_module()` is triggered for all physical surfaces including the beampipe. Do you know a way to exclude the passive surfaces?